### ISSUE TITLE
[API Proposal]: Allow for setting of the `Elapsed` property in the `System.Diagnostics.Stopwatch` class. 

### ISSUE THREAD

**Issue by FilipToth** [dotnet/runtime#56817]
## Background and Motivation
I've noticed while working with the `Stopwatch` class that you can't just set the `Elapsed` (`ElapsedMiliseconds` and `ElapsedTicks` too). It would be useful for working on projects that require precise time controls (eg. Audio Editors). I do understand that the stopwatch class was mostly meant for use in diagnostics (Hence the namespace.) not actual production code, but I still think there would be a ton of benefits even in diagnostic implementations.

## API Proposal
```cs
public TimeSpan Elapsed {
    get { return new TimeSpan( GetElapsedDateTimeTicks()); }
    set { elapsed = ...; Elapsed = new TimeSpan(0, 0, ...) }
}
```

## API Usage
```cs
private void SetTime(long ms)
{
    _stopwatch.Elapsed = new TimeSpan(0, 0, ms);
    _bar.Left = CalculateLeftPosFromMs(ms);
}
```
This was a quick snipped from a super-quick audio editor project, I've needed to build.

## Risks
None that I can think of, if properly implemented. This would not be a breaking change (at least, I don't think so).

**Comment by GrabYourPitchforks** [comment]
> \[Risks\] None that I can think of, if properly implemented.

One potential risk is that it's not possible for `Stopwatch.Elapsed` to express every possible `TimeSpan`. This is due to the multiplication that takes place here:

https://github.com/dotnet/runtime/blob/79c4144f174754bf4fa18827a292549214f86581/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Stopwatch.cs#L133-L139

The *s_tickFrequency* field describes the smallest quantum of measurable time within the `Stopwatch` instance. If you generate a `TimeSpan` value whose tick count is not a whole multiple of this quantum, then you call `Stopwatch.set_Elapsed` with this time span, then `Stopwatch.get_Elapsed` will return a different value. This may be unexpected by callers and could manifest as undesirable behavior at the call site.

**Comment by ryank425** [comment]
> > [Risks] None that I can think of, if properly implemented.
> 
> One potential risk is that it's not possible for `Stopwatch.Elapsed` to express every possible `TimeSpan`. This is due to the multiplication that takes place here:
> 
> https://github.com/dotnet/runtime/blob/79c4144f174754bf4fa18827a292549214f86581/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Stopwatch.cs#L133-L139
> 
> The _s_tickFrequency_ field describes the smallest quantum of measurable time within the `Stopwatch` instance. If you generate a `TimeSpan` value whose tick count is not a whole multiple of this quantum, then you call `Stopwatch.set_Elapsed` with this time span, then `Stopwatch.get_Elapsed` will return a different value. This may be unexpected by callers and could manifest as undesirable behavior at the call site.

I think this problem can be resolved by making the API call to round up to the multiples of tickFrequency that device can support.

That would make `private void SetTime(long ms)` actually be `private long SetTime(long ms)` to return the rounded time as it is impossible to get that level of accuracy with given constraints in this environment anyway.

**Comment by GrabYourPitchforks** [comment]
> I think this problem can be resolved by making the API call to round up to the multiples of tickFrequency that device can support.

Yes, you could round it. But this means that the caller needs to readjust their zero baseline from what they might otherwise expect it to be. For example, if I set `myStopwatch.Elapsed = TimeSpan.FromSeconds(1);`, and if I want to know if the `Stopwatch` instance has ever been started after I set the _Elapsed_ property, I need to subtract the new _Elapsed_ property value from 1 second + ε, not simply 1 second. Property values silently changing to something other than what the caller set is one way for an application to encounter bugs, particularly when they start receiving non-zero (or negative!) values in calculations that they'd reasonably expect to produce zero.

**Comment by ryank425** [comment]
> > I think this problem can be resolved by making the API call to round up to the multiples of tickFrequency that device can support.
> 
> Yes, you could round it. But this means that the caller needs to readjust their zero baseline from what they might otherwise expect it to be. For example, if I set `myStopwatch.Elapsed = TimeSpan.FromSeconds(1);`, and if I want to know if the `Stopwatch` instance has ever been started after I set the _Elapsed_ property, I need to subtract the new _Elapsed_ property value from 1 second + ε, not simply 1 second. Property values silently changing to something other than what the caller set is one way for an application to encounter bugs, particularly when they start receiving non-zero (or negative!) values in calculations that they'd reasonably expect to produce zero.

I agree with your points. As for the OP's case, I think you can have still use stopwatch for timing as I am not aware of other mechanism to get high resolution (<1us) time stamp that can be used for time-interval measurements. Just set another variable to hold your custom Elapsed, and add the measurement from Stopwatch as this would be the closest accurate measurement you can get from OS.

**Comment by FilipToth** [comment]
Ok, we could still have a set accessor in the `ElapsedMilliseconds` property and the `ElapsedTicks` property (if we expose the `TicksPerMillisecond` field). As for negative numbers, we just don't set the property if the value is negative.

**Comment by josalem** [comment]
"Tick" is an overloaded term here. There are "DateTime Ticks" which are 1 tick per 100ns and then there are "System Clock Ticks" which have the frequency returned by `Stopwatch.Frequency`.

[`TicksPerMillisecond` is a `const` value of `10000`](https://source.dot.net/#System.Private.CoreLib/Stopwatch.cs,12) for converting the underlying system frequency (the value returned by `Stoopwatch.Frequency`) to `DateTime` ticks (100ns ticks). See The implementation of [`get_Elapsed`](https://source.dot.net/#System.Private.CoreLib/Stopwatch.cs,96) for details. You can do the calculations yourself if you use `Stopwatch.GetTimestamp` and `Stopwatch.Frequency`.

I think @ryank425 is correct in suggesting that if you want to get "elapsed since timestamp" you should track the timestamp separately.

If you're using `Stopwatch` to track time deltas, you should make sure to convert your elapsed value to the same clock used by your time interval, e.g., your audio editor workspace. You'll want to use the `Stopwatch.{Start|Stop}` methods with `Stopwatch.ElapsedTicks` and `Stopwatch.Frequency` to convert to the correct clock for your use case. A simple wrapper around the `Stopwatch` class should give you want you want.

**Comment by FilipToth** [comment]
> A simple wrapper around the Stopwatch class should give you want you want.
Yes, that is exactly what I did in my own code before submitting the issue, I was just wondering about if this could be a positive change to the .NET runtime.

**Comment by tommcdon** [comment]
Moving older issues to the Future milestone as it is unlikely we will have time to address this in the .NET 7 timeframe.  Please feel free to move this issue back to .NET 7 if there is active work to address the feature request.
